### PR TITLE
Precint crud

### PR DESCRIPTION
**Pull request by andrei1489**
+ created docker-compose config to allow developers to work on this without installing wamp/lamp etc
+ created crud functionality for precincts with swagger documentation
+ created an import endpoint that takes the provided xlsx/csv and imports it
* refactored the seeder to use the same class

### CODE REVIEW

File: app/Helpers/CsvHandler.php [hunk 1]
```diff
@@ -0,0 +1,35 @@
+<?php
```
Comment (claudiunicolaa) [inline app/Helpers/CsvHandler.php:1]: please, use proper indentation for functions content :)

File: app/Helpers/CsvHandler.php [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+	protected $delimiter = ',';
+
+	public static function convertToArray($path)
+	{
+		if(!file_exists($path) || !is_readable($path)) {
+			return FALSE;
+		}
+
+
+        $data = array();
```
Comment (claudiunicolaa) [inline app/Helpers/CsvHandler.php]: `[]` instead of `array()`

File: app/Helpers/CsvHandler.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+		{
+            $data = self::convertFileToArray($handle);
+			fclose($handle);
+		}
+
+		return $data;
+	}
+
+	public static function convertFileToArray($file){
+        $data = array();
```
Comment (claudiunicolaa) [inline app/Helpers/CsvHandler.php]: [] instead of array()

File: app/Helpers/PrecinctImporter.php [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+namespace App\Helpers;
+use Akeneo\Component\SpreadsheetParser\SpreadsheetParser;
+use App\City;
+use App\Precinct;
+use Illuminate\Support\Facades\File;
+
+class PrecinctImporter
+{
+
+    public function importFromFile(\Symfony\Component\HttpFoundation\File\File $file, bool $deleteFileAfter)
```
Comment (claudiunicolaa) [inline app/Helpers/PrecinctImporter.php]: import the class `\Symfony\Component\HttpFoundation\File\File` and use the simplified FQN (`File`) (for keeping the usages of type hints as in rest of the project)

File: app/Helpers/PrecinctImporter.php [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            $data = $this->readFromXLSX($file);
+        }
+
+        $this->importPrecinctsFromArray($data);
+        if ($deleteFileAfter) {
+            File::delete($file);
+        }
+    }
+
+    private function readFromCSV(\Symfony\Component\HttpFoundation\File\File $file)
```
Comment (claudiunicolaa) [inline app/Helpers/PrecinctImporter.php]: import the class `\Symfony\Component\HttpFoundation\File\File` and use the simplified FQN (`File`) (for keeping the usages of type hints as in rest of the project)

File: app/Helpers/PrecinctImporter.php [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+                    'headquarter' => $rawPrecinctData[4],
+                    'address' => $rawPrecinctData[5],
+                ];
+            }
+        }
+        return $precinctData;
+
+    }
+
+    private function readFromXLSX($file)
```
Comment (claudiunicolaa) [inline app/Helpers/PrecinctImporter.php]: type hint for `$file`

File: app/Helpers/PrecinctImporter.php [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+                        'address' => $rowData[6]
+                    ];
+                }
+            }
+        }
+        return $data;
+
+    }
+
+    private function importPrecinctsFromArray($data)
```
Comment (claudiunicolaa) [inline app/Helpers/PrecinctImporter.php]: type hint for $data

File: app/Helpers/PrecinctImporter.php [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    {
+        foreach ($data as $rawPrecinctData) {
+            $precinct = new Precinct($rawPrecinctData);
+            $precinct->save();
+        }
+    }
+    private function getCityId($cityName)
```
Comment (claudiunicolaa) [inline app/Helpers/PrecinctImporter.php]: type hint for $cityName

File: app/Helpers/PrecinctImporter.php (hunk 4 above)
Comment (andrei1489) [reply app/Helpers/PrecinctImporter.php]: there is a name collision between `Illuminate\Support\Facades\File` and `\Symfony\Component\HttpFoundation\File\File`. I can alias the Facades one to Filesystem ( to have something like Filesystem::delete() and use the simplified File for type hinting.

Comment (andrei1489) [reply app/Helpers/PrecinctImporter.php]: Decided to move away from framework specific classes for file and used SplFileObject

**Comment by claudiunicolaa** [comment]
:twisted_rightwards_arrows: From my point of view can be merged. @andrei1489 it will be nice if you can add into readme the instructions for using docker?

**Comment by andrei1489** [comment]
@claudiunicolaa added readme section on docker

**Comment by andrei1489** [comment]
@claudiunicolaa at the moment there is no way of preventing duplicates. Should I open a new issue for this? or can you provide how a unique precinct might be determine (composite key by city/precinct_no) and I will incorporate it in this task?

**Comment by andrei1489** [comment]
as per the discussion @claudiunicolaa I will do the following:
- create a migration to create a composite key for precincts ( precinct_no and city_id)
- during the import, if a precinct already exists, update the precinct
- during the import, if a precinct does not have valid info, skip it and continue the import

**Comment by andrei1489** [comment]
import feature now works with json, but the expected json has to be in the diaspora.json format. Unfortunately due to the lack of control on the format of the documents regarding precincts, the import endpoint only works like this:
- xlsx, json and csv file have to be the same format as the ones in resources
- json file will always be treated as diaspora

File: app/Api/V1/Controllers/PrecinctController.php [hunk 9]
```diff
@@ -2,30 +2,322 @@
...
+     *     @SWG\Schema(ref="#/definitions/Precinct")
+     *   ),
+     *   @SWG\Response(response=400,  description="Invalid Precinct"),
+     *   @SWG\Response(response=404,  description="Invalid Precinct"),
+     *   @SWG\Response(response=409,  description="Duplicate Precinct")
+     * )
+     */
+    public function update(Request $request, int $id)
+    {
+        $data = $request->only(array_keys($this->validationRules));
```
Comment (RaduCStefanescu) [inline app/Api/V1/Controllers/PrecinctController.php:146]: This looks like DRY, maybe you can consider creating a validation method.

**Comment by RaduCStefanescu** [comment]
> as per the discussion @claudiunicolaa I will do the following:
> 
> * create a migration to create a composite key for precincts ( precinct_no and city_id)
> * during the import, if a precinct already exists, update the precinct
> * during the import, if a precinct does not have valid info, skip it and continue the import

Hi @andrei1489 I think right now this is the best way to go. we will maintain the file format for precincts as the one already existing in resources, and indeed, the combination of precinct number and city id should be unique. Actually, the uniqueness is at county id level, as each county has a number of precincts, numbered starting from 1.

**Comment by RaduCStefanescu** [comment]
Thank you for the PR @andrei1489 merging it right now.

**Pull request merged**